kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs

```diff
@@ -1991,30 +1991,42 @@ impl<'tcx> GotocCtx<'tcx> {
         let x = fargs.remove(0);
         let y = fargs.remove(0);
 
-        // if(same_object(x, y)) {
-        //   assert(x + 1 <= y || y + 1 <= x);
-        //   assume(x + 1 <= y || y + 1 <= x);
-        // }
-        let one = Expr::int_constant(1, Type::c_int());
-        let non_overlapping =
-            x.clone().plus(one.clone()).le(y.clone()).or(y.clone().plus(one.clone()).le(x.clone()));
-        let non_overlapping_check = self.codegen_assert_assume(
-            non_overlapping,
-            PropertyClass::SafetyCheck,
-            "memory regions pointed to by `x` and `y` must not overlap",
-            loc,
-        );
-        let non_overlapping_stmt =
-            Stmt::if_then_else(x.clone().same_object(y.clone()), non_overlapping_check, None, loc);
-
-        // T t = *y; *y = *x; *x = t;
         let deref_y = y.clone().dereference();
-        let (temp_var, assign_to_t) =
-            self.decl_temp_variable(deref_y.typ().clone(), Some(deref_y), loc);
-        let assign_to_y = y.dereference().assign(x.clone().dereference(), loc);
-        let assign_to_x = x.dereference().assign(temp_var, loc);
+        if deref_y.typ().sizeof(&self.symbol_table) == 0 {
```

Comment:
Should we use `is_zst(..)` here?

Reply:
For all that I know `is_zst(..)` exists for MIR types only, and what we have is a pointer type where the ZST would be the pointed-to type. If anyone can teach me how to get the pointed-to type at MIR level then I'd happily use `is_zst(..)` (and we should then also consider doing so in other places).

Reply:
There are actually a few `is_zst*` methods defined in `kani-compiler`. I think you can directly use [this one](https://github.com/model-checking/kani/blob/ca110f7b2c1d7cf219595863ef355713b306dc28/kani-compiler/src/codegen_cprover_gotoc/codegen/ty_stable.rs#L42-L44) in your code.

Reply:
https://github.com/model-checking/kani/blob/ca110f7b2c1d7cf219595863ef355713b306dc28/kani-compiler/src/codegen_cprover_gotoc/codegen/ty_stable.rs#L125

Reply:
Thank you!